### PR TITLE
aio.core: Add async functional functions

### PR DESCRIPTION
**Pull request by phlax**
- async_list
- async_set
- async_map
- maybe_coro
- maybe_awaitable

Signed-off-by: Ryan Northey <ryan@synca.io>

### CODE REVIEW

File: aio.core/aio/core/functional/collections.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+from typing import Any, AsyncIterable, Callable, List, Optional, Set
+
+
+async def async_set(
+        iterable: AsyncIterable,
+        predicate: Optional[Callable[[Any], bool]] = None,
+        result: Optional[Callable[[Any], Any]] = None) -> Set:
+    """Create a set from the results of an async generator."""
+    results = set()
+    result = result or (lambda path: path)
```
Comment (phlax) [inline aio.core/aio/core/functional/collections.py:11]: `path` -> `item` - same below

File: aio.core/aio/core/functional/process.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+
+from concurrent import futures
+from typing import AsyncIterator, Callable, Iterable
+
+
+async def async_map(
+        fun: Callable,
+        iterable: Iterable,
+        fork: bool = False) -> AsyncIterator:
+    """Asynchronously map sync fun with an iterable, yielding results as
```
Comment (kfaseela) [inline aio.core/aio/core/functional/process.py:10]: expand fun? :)

File: aio.core/aio/core/functional/process.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    available.
+
+    By default a `ThreadPoolExecutor` will be used, setting `fork` to
+    `True` will instead use a `ProcessPoolExecutor` to fork the fun to
```
Comment (kfaseela) [inline aio.core/aio/core/functional/process.py:14]: fun

Comment (phlax) [reply aio.core/aio/core/functional/process.py:14]: i guess im just trying to put some fun into coding - ill remove/expand 8/

**Comment by phlax** [comment]
WIPping temporarily while add functionality to allow async predicates/result handlers

**Comment by phlax** [comment]
v temp reWIPping while i resolve some typecheck issues...

**Review by kfaseela — APPROVED** [review]

**Pull request merged**